src/main/java/frc/robot/swerve/ModuleIOReal.java

```diff
@@ -0,0 +1,181 @@
+        inputs.driveMotorPosition = driveMotor.getRotorPosition().getValue();
+        inputs.driveMotorVelocity = getVelocity();
+        inputs.driveMotorVelocitySetpoint = driveMotorVelocitySetpoint;
+
+        inputs.angleMotorSupplyCurrent = angleMotor.getSupplyCurrent().getValue();
+        inputs.angleMotorStatorCurrent = angleMotor.getStatorCurrent().getValue();
+        angleSupplyChargeUsedCoulomb.update(inputs.angleMotorSupplyCurrent);
+        inputs.angleMotorSupplyCurrentOverTime = angleSupplyChargeUsedCoulomb.get();
+        angleStatorChargeUsedCoulomb.update(inputs.angleMotorStatorCurrent);
+        inputs.angleMotorStatorCurrentOverTime = angleStatorChargeUsedCoulomb.get();
+        inputs.angleMotorPosition = angleMotor.getRotorPosition().getValue();
+        inputs.angleMotorVelocity = angleMotor.getVelocity().getValue();
+
+        inputs.angle = getAngle();
+        currentAngle = inputs.angle;
+
+        inputs.angleSetpoint = angleSetpoint;
+
+        inputs.moduleDistance = getModulePosition().distanceMeters;
+    }
+
+    @Override
+    public double getAngle() {
+        return AngleUtil.normalize(Units.rotationsToRadians(angleMotor.getPosition().getValue()));
+    }
+
+    @Override
+    public void setAngle(double angle) {
+        angleSetpoint = AngleUtil.normalize(angle);
+        Rotation2d error = new Rotation2d(angle).minus(new Rotation2d(currentAngle));
+        angleMotor.setControl(
+                new MotionMagicVoltage(angleMotor.getPosition().getValue() + error.getRotations())
+                        .withEnableFOC(true));
+    }
+
+    @Override
+    public double getVelocity() {
+        return driveMotor.getVelocity().getValue();
+    }
+
+    @Override
+    public void setVelocity(double velocity) {
+        var angleError = new Rotation2d(angleSetpoint).minus(new Rotation2d(currentAngle));
+        velocity *= angleError.getCos();
+
+        driveMotorVelocitySetpoint = velocity;
+        driveMotor.setControl(
+                new VelocityVoltage(utils.units.Units.metersPerSecondToRps(
+                        velocity,
+                SwerveConstants.WHEEL_DIAMETER / 2
+        )).withEnableFOC(true));
+    }
+
+    @Override
+    public SwerveModulePosition getModulePosition() {
+        return new SwerveModulePosition(
+                utils.units.Units.rpsToMetersPerSecond(
+                        driveMotor.getPosition().getValue(),
+                        SwerveConstants.WHEEL_DIAMETER / 2
+                ),
+                new Rotation2d(getAngle())
+        );
+    }
+
+    @Override
+    public void updateOffset(double offset) {
+        angleMotor.setPosition(
+                ((encoder.getAbsolutePosition() - offset) * 2048) / SwerveConstants.ANGLE_REDUCTION);
+    }
+
+    @Override
+    public void neutralOutput() {
+        driveMotor.setControl(new NeutralOut());
+        angleMotor.setControl(new NeutralOut());
```

Comment:
Control request should be cached.

Reply:
What do you mean?

Reply:
From the [Phoenix 6 documentation](https://pro.docs.ctr-electronics.com/en/latest/docs/api-reference/api-usage/control-requests.html#modifying-a-control-request):
> Java users should reuse control requests to prevent excessive invocation of the Garbage Collector.

You can create the request object as a class member and change the value as needed with [`withPosition()`](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/controls/PositionVoltage.html#withPosition(double)) (this doesn't create a new object but changes the same one).

Reply:
But the neutral output is a different control mode, I can't use the method.

Reply:
The second sentence does only apply to control requests with variable parameters. You can still create a `NeutralOut` control request as a class member and reuse it.

I also noticed yesterday that there is a [`TalonFX.stopMotor()`](https://api.ctr-electronics.com/phoenix6/release/java/com/ctre/phoenix6/hardware/TalonFX.html#stopMotor()] (credit to @GaiaZano05) which looks cleaner.